use google/cpu_features

alexcohn opened the pull request:
NDK/cpufeatures has been [deprecated](https://developer.android.com/ndk/guides/cpu-features) in favor of [google/cpu_features](/google/cpu_features). As it's a [drop-in replacement](https://github.com/google/cpu_features/tree/master/ndk_compat) for NDK/cpufeatures, we can switch to it at (relatively) low cost.

tanersener commented:
The situation about `cpufeatures` is more complicated than it looks. I wanted to use `android/cpufeatures` before. The  problem is `libvpx` library. `libvpx` depends on `cpufeatures` and requires `cpufeatures` to be built before the NDK part of the build process. Because of that, I'm calling `build_cpufeatures` function in `main-android.sh` to build `cpufeatures` before compiling external libraries. 

Anyway, when I was testing using `android/cpufeatures` module I had different symbol errors on linking. So, I had to cancel that change. I'm not sure if that change is still in the git history. Did you test this change with `--enable-libvpx` ?

alexcohn commented:
Yes, it builds with `--enable-libvpx` for all all 4 architectures. The difference, probably, is that I tuned the flags for NDK 20 and higher (in an earlier commit).

tanersener commented:
> is that I tuned the flags for NDK 20 and higher (in an earlier commit).

Can you show me that commit. I don't see any big changes regarding the build flags?

tanersener commented:
Tested changes in this branch. Built all 5 architectures and created an LTS archive. In terms of compilation and linking I didn't receive any errors. 

The main problem is `.aar` file created does not include `libcpufeatures.so` inside. So, If I try to load the library I receive the following error. Did you managed to load the library created in this branch?

```
2020-04-07 01:45:53.481 8213-8213/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.arthenica.mobileffmpeg.test, PID: 8213
    java.lang.UnsatisfiedLinkError: dlopen failed: library "libcpufeatures.so" not found
        at java.lang.Runtime.loadLibrary0(Runtime.java:977)
        at java.lang.System.loadLibrary(System.java:1530)
        at com.arthenica.mobileffmpeg.Config.<clinit>(Config.java:131)
        at com.arthenica.mobileffmpeg.Config.setFontDirectory(Config.java:337)
        at com.arthenica.mobileffmpeg.test.MainActivity.registerAppFont(MainActivity.java:168)
        at com.arthenica.mobileffmpeg.test.MainActivity.onCreate(MainActivity.java:123)
```

alexcohn commented:
> `.aar` file created does not include `libcpufeatures.so` inside.

Oh, I see. NDK normally builds libcpufeatures as static library. You create two shared libraries: one before you configure ffmpeg with libvpx (to satisfy the linker), and another one in android/app (as part of jni/Android.mk). Note that openh264 chose a different path: they compile cpu-features.c into their static lib.

In the end, you only need to compile cpu-features.c once, and let everybody use the same `android_getCpuFeatures()`. Also, Google has [deprecated](https://developer.android.com/ndk/guides/cpu-features#the_ndk_cpufeatures_library_deprecated) **sources/android/cpufeatures**.

I will try to check what it takes to use the [Google updated library](https://github.com/google/cpu_features) and update this PR.

I guess that this means

tanersener commented:
Okay Alex, thanks.

alexcohn commented:
OK, one puzzle is now settled: as long as you build **libvpx** as a static library, it doesn't need **cpu_features** to build. And as long as it relies on **cpu-features.h**, the google new implementation can replace the one bundled into NDK, and it can be built as shared library.

So, **cpufeatures** does not need special treatment at all, only make sure it's always enabled.

The one piece that is missing now is that **openh264** compiles **cpu-features.c** from NDK sources (it also renames all the functions to `wels_XXX`). Working around this requires patching `src/openh264/build/platform-android.mk`, and I am afraid this is out of scope of this project. I will propose this change to https://github.com/cisco/openh264.

BTW, speaking of **openh264**: the library doesn't actually depend on libc++, and this is reflected in this PR. 

While adding support for **google/cpu_features** to **Android.mk** files, I also cleaned these up just a little bit. I have few more improvements for NDK build process, but they will be submitted separately.

You will notice that **cpu_features** Library is built out-of-tree, into `android/build` directory. This was a completely arbitrary choice, so this can be easily changed, if you prefer a different location. Actually, many other libraries support out-of-tree, e.g. **libvpx**. Using this can reduce rebuild time significantly.

Another note: I added **cpu_features** to the end of the list of *external* libraries. This was also an arbitrary choice, and the only incentive was to reduce amount of diff in the numbered lists of libraries.

tanersener commented:
Thanks for working on this feature. 
Did you try to run/load created libraries? I can't load them from `MobileFFmpeg` Java Api. Created `openh264` enabled libraries. When I load them in test-app, application crashes on startup with errors about `libc++_shared.so`.

alexcohn commented:
Yes, I ran the test app. I will check it again: maybe my build tree was not clean.

alexcohn commented:
I ran the minimal build and the full build, both were fine. Now I tried `./android.sh  --enable-openh264` and yes, this combination crashed. Further checks revealed that **openh264.pc** lists **c++** lib dependency which is not really necessary. I have removed it, and resolved this bug.

Possibly, link to a C++ runtime was necessary for some earlier NDK release, maybe some functions that are inline today needed runtime support with **stlport**. Even more probably, this flag is relevant for other platforms that **openh264** can be built for.

tanersener commented:
Well, this new version works on Main builds but still fails for LTS ones. Build is successful, and some changes are needed inside the static blocks of `Config` & `AbiDetect` classes. But loading the native libraries still fails due to some undefined symbols. Not sure if it is possible to resolve them without adding `c++` dependency.

alexcohn commented:
OK, I will check what happens on LTS.

On Sun, May 17, 2020 at 1:09 AM Taner Sener <notifications@github.com>
wrote:

> Well, this new version works on Main builds but still fails for LTS ones.
> Build is successful, and some changes are needed inside the static blocks
> of Config & AbiDetect classes. But loading the native libraries still
> fails due to some undefined symbols. Not sure if it is possible to resolve
> them without adding c++ dependency.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tanersener/mobile-ffmpeg/pull/375#issuecomment-629712303>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTZM3S2ODJ3V22N7TITB3RR4FI7ANCNFSM4MCOOO3Q>
> .
>

alexcohn commented:
That's true. 

So I simply reverted the change that excluded **c++_shared** for **openh264** (after all, this change has nothing to do with the cpu-features feature). I don't think that it's worth to add more intelligent decision tree here that will choose between LTS and regular build. The extra library makes very little difference anyway.

I also added the changes to static constructor blocks that you mentioned above, and I decided to make the code simpler (at the cost of a try&hellip;catch).

tanersener commented:
Thanks. There are a few minor issues that needs to be addressed during the code review. But before resolving them, I need to explain my main concern about this PR: submodule.

I know that the repository is not small and it is getting bigger and bigger each day. Including the source code of each library does not help much in terms of repo size. But I made that decision long time ago and I want to continue that way. Explained it here #84. The main idea is not depending to another project/module. Source code of those projects/modules/submodules can change and this can break the `mobile-ffmpeg` builds. I simply don't want that, so I add the source code of external libraries to the repository. Not the best solution but brings the stability for this project.

In terms of this PR, you don't have to change anything. I can add the source code of `cpu_features` and remove the submodule later in another commit if that works for you.

alexcohn commented:
I carefully read the [tarball discussion](https://github.com/tanersener/mobile-ffmpeg/issues/84) and I fully understand that there can be no easy split-strategy-fit-all in the case of so many dependencies, each with their own release strategy, version control and deployment technologies. To keep the discussion focused, I have posted an answer in the same thread.

tanersener commented on android/app/src/main/java/com/arthenica/mobileffmpeg/Config.java:
```diff
@@ -84,12 +84,13 @@
             }
-            System.loadLibrary("cpufeatures");
+            catch (UnsatisfiedLinkError ex) {
+                Log.i(Config.TAG, "libc++_shared.so fialed to load, try to proceed without it");
```
Can we revert this change? We know which libraries depend on `c++`. We even have a wiki page about it. Not using that information and printing an unnecessary log line doesn't make sense to me.

alexcohn replied:
We can simply skip the log line, as well. I thought that reducing complex decision here would make maintenance easier: there is only one place in code that decides to package C++ runtime, and the rest follow that decision. This way the decision may be refined. Like, as I found, **openh264** only needs C++ runtime in LTS build.

But if you are uncomfortable with this, I will happily revert this change.

alexcohn replied:
So, in the meanwhile, I simply revert this change.

tanersener commented on build/android-cpu_features.sh:
```diff
@@ -0,0 +1,69 @@
+echo ${cmake}
+}
+
+create_cpu_features_package_config() {
```
All custom package config functions are created inside `android-common.sh`. Is there a reason to define it here?

alexcohn replied:
I thought that handling all **cpu_features** related logic in single script would be cleaner. I don't see any advantage in keeping it in `android-common.sh`.

tanersener replied:
Well, it is a design decision taken to keep library build scripts short and easy to understand. You can see that none of those scripts include function definitions.

alexcohn replied:
as you wish

tanersener commented on build/android-cpu_features.sh:
```diff
@@ -0,0 +1,69 @@
+
+PREFIX=${BASEDIR}/prebuilt/android-$(get_target_build)/${LIB_NAME}
+
+android_cmake() {
```
Using Android's cmake makes sense. What do you think about defining it globally inside `set_toolchain_clang_paths`? So, other libraries can use it too.

alexcohn replied:
Definitely, this can be moved to `android-common.sh`.

alexcohn replied:
What if I put a new 'function' **android_cmake()** into `android-common.sh`? 

This will handle running `cmake` with NDK toolchain and necessary arguments. You still can add more arguments to customize the build, e.g. `BUILD_SHARED_LIBS=ON`.

This same function can be used for all libraries that currently build with `cmake`, and hopefully, also help to switch other libraries, ones that support `cmake`, to avoid the slow `./configure`.

tanersener commented on android/jni/cpu_features/Android.mk:
```diff
@@ -0,0 +1,8 @@
+include $(CLEAR_VARS)
+LOCAL_ARM_MODE := $(MY_ARM_MODE)
+LOCAL_MODULE := cpu_features
+LOCAL_SRC_FILES := libndk_compat.so
```
Historically we used `libcpufeatures.so` as `cpu-features` library name. Do you know if it is possible to continue using that name?

alexcohn replied:
Yes it is possible, but I don't see why the true name is worse. I mean, we import a well-known library from a quite respectable source. Renaming it may only cause confusion and result in that some people will end with two copies of the library (if they use it their NDK code).

Note that in the long run, it is possible to gradually switch to the more detailed **libcpu_features.so**, and also, probably, use it for iOS.

tanersener replied:
The idea is simple. Name of the library is `cpu-features` and the name of the library file is `libndk_compat.so`. They look like two different things. I just expect to have a filename which will include the name of the library. So, people looking inside the .apk can understand what it is.

alexcohn replied:
OK then, you could name the library `ndk-compat`. Or hide it altogether from the list, consider it an implementation detail, as you don't have a separate entry for `mobileffmpeg_abidetect` or `androidltssupport`.

Alternatively, you can use it as a static library, then the name of `libndk_compat.a` can be harmlessly changed to `libcpufeatures.a`.

Finally, the long-term solution would be to use the modern `libcpu_features.so` that is part of the same Google repo. Let me explain.

Google developed this cross-platform library that handles more CPU features than the small one that is (still) included in NDK, but its API is different. They maintain the `libndk_compat.so` library in the same GitHub project, which is Android-only and probably less advanced, but replaces seamlessly the library that the Android developers are used to.

tanersener replied:
`mobile-ffmpeg` can switch to main `cpu-features` library, even today. However, `libvpx` still requires NDK compatible one. So, we have to wait for that.

Generating a static library can be a solution. But, it is not the ideal choice. Because linking the same static library from multiple libraries creates complications about the symbols. I'm not % 100 sure that it will be possible to resolve them and build a running version.

Hiding the library name is possible. I'm already doing this in current version of the scripts. But this is not my main concern. My main concern is the file name. `ndk_compat` is not telling me anything about `cpu-features`.

By the way, I can solve it in my way if you don't want to do anything about it.

alexcohn replied:
I find no complications when I use **ndk_compat** as a static library. The change is in a separate commit, and can be easily reverted if you wish.

tanersener replied:
I guess you're right about the static version of the library. Two different libraries are linked with it. So, it shouldn't be a problem.

tanersener commented on build/android-common.sh:
```diff
@@ -63,8 +63,9 @@ get_library_name() {
         44) echo "libsamplerate" ;;
-        45) echo "android-zlib" ;;
-        46) echo "android-media-codec" ;;
+        45) echo "cpu_features" ;;
```
We used `-` (dash) instead of `_` (underscore) in all external library names. Can we continue following that approach?

alexcohn replied:
Yes, here there is less reason to stick to the original name. Especially if this library is enabled automatically.

alexcohn replied:
done

tanersener commented on android/jni/Android.mk:
```diff
@@ -21,40 +21,35 @@ ifeq ($(TARGET_ARCH_ABI), x86_64)
 
 $(call import-module, ffmpeg)
+$(call import-module, cpu_features)
+$(call import-module, cxx-stl/llvm-libc++)
```
Hmm, what is the purpose of importing `cxx-stl/llvm-libc++` module here?

alexcohn replied:
To compensate for craziness of Android NDK. If they sniff that the project uses C++, they add the same line `$(call import-module, cxx-stl/llvm-libc++)` automatically for you. But in **mobile-ffmpeg**, all files are C. So they decide not to add it, and simple `LOCAL_SHARED_LIBRARIES += c++_shared` fails because they don't know where to find it.

tanersener replied:
Okay, but I didn't had to add it before. We shouldn't be needing this.

alexcohn replied:
You're right, this import is done automatically in the new NDK.

I am removing the offending line.

alexcohn commented on android.sh:
```diff
@@ -53,15 +53,18 @@ LIBRARY_EXPAT=41
 LIBRARY_LIBSAMPLERATE=44
-LIBRARY_ZLIB=45
-LIBRARY_MEDIA_CODEC=46
+LIBRARY_CPU_FEATURES=45
```
Quite possible. Actually, my first try was to set it at **0**, this also works. The drawback is that this changes all the numbers.

alexcohn replied:
BTW, why are some loops starting at 0, while others at 1 (and use `($library - 1)` inside)?

tanersener replied:
I remember that I had to start loops from 1 due to a limitation of bash, in the early versions of the build scripts. But I don't remember what kind of issue that was. 

It looks like it is possible to start from 0 in some of those loops. We can update them.

alexcohn commented on android.sh:
```diff
@@ -53,15 +53,18 @@ LIBRARY_EXPAT=41
 # ENABLE LIBRARIES
 ENABLED_LIBRARIES=(0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0)
 
+ENABLED_LIBRARIES[LIBRARY_CPU_FEATURES]=1
```
That's true, I considered this, but in the end I decided that such thing should be more explicit. But as you wish…

If we move cpu_features to the end, this will look cleaner, anyways.

alexcohn commented on android.sh:
```diff
@@ -262,6 +265,9 @@ set_library() {
   chromaprint)
     ENABLED_LIBRARIES[LIBRARY_CHROMAPRINT]=$2
     ;;
+  cpu_features)
```
I thought this could help in case somebody wants `-rebuild-cpu_features` for any reason.

tanersener replied:
You're right, we need to support rebuilding of `cpu-features`. But that functionality is not implemented here.

alexcohn replied:
The actual reason this was needed here was that the loop for `--full` included this library, too. Now that it is the last (number 47) and always enabled, this block can be removed.

alexcohn commented on android.sh:
```diff
@@ -785,7 +789,7 @@ for run_arch in {0..4}; do
 
     # CLEAR FLAGS
-    for library in {1..47}; do
+    for library in {1..46}; do
```
Yes, my fault.

alexcohn commented on android/app/src/main/java/com/arthenica/mobileffmpeg/AbiDetect.java:
```diff
@@ -35,7 +36,16 @@
         if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
-            System.loadLibrary("cpufeatures");
+            try {
+                System.loadLibrary("c++_shared");
```
This was introduced to allow loading **libavcodec**. There was some unresolved symbol before (in case of LTS build), but I don't remember exactly how it happened. Anyways, this is closely related to https://github.com/tanersener/mobile-ffmpeg/pull/375#discussion_r427548560.

alexcohn commented on android/jni/Android.mk:
```diff
@@ -21,40 +21,35 @@ ifeq ($(TARGET_ARCH_ABI), x86_64)
+LOCAL_CFLAGS := -Wall -Wextra -Werror -Wno-unused-parameter -DMOBILE_FFMPEG_${MY_ARCH_FLAGS}
 LOCAL_LDLIBS := -llog -lz -landroid
-LOCAL_SHARED_LIBRARIES := cpufeatures
+LOCAL_SHARED_LIBRARIES := cpu_features libavcodec libavutil
```
There was some unresolved symbol that could be found in **libavcodec** (in case of LTS build), but I don't remember exactly how it happened. I will check again.

tanersener replied:
Dependencies given here define which libraries should be loaded in `AbiDetect` class. It is important to keep it short and not depend any `libabxx` library.

alexcohn replied:
The problem is that `mobileffmpeg_abidetect.c` includes `fftools_ffmpeg.h` which includes quite a few headers from **libavutil** and **libavcodec**. The result requires [this line](https://github.com/tanersener/mobile-ffmpeg/pull/375/commits/de7f3fcef7aaf58497fdc68ee64bed614c249f9b#diff-983cf7060ee7932325a5473d56993529R3), which could be expressed by a cleaner (IMHO)

    LOCAL_SHARED_LIBRARIES := cpu-features libavutil

(`libavcodec` proved to be unnecessary here)

tanersener replied:
`mobileffmpeg_abidetect.c` includes `fftools_ffmpeg.h` to access `FFMPEG_CONFIGURATION`. But we don't need to depend `libavutil` for that. Just having the include is enough. I think you doing this in the latest version.

alexcohn replied:
Yes, that's what I do, but I don't like the ugly path that I put there. You don't like a fake dependency even more, so let it be it.

alexcohn commented:
I have run a full search for use of `cpu-features.h` in all libraries under `/src`:

- `libaom/aom_ports/arm_cpudetect.c`
- `libpng/contrib/arm-neon/android-ndk.c`
- `libvpx/vpx_ports/arm_cpudetect.c`
- `libwebp/src/dsp/cpu.c`
- `openh264/codec/common/src/cpu.cpp`
- `openh264/codec/common/src/WelsThreadLib.cpp`

This PR only takes care of **libvpx**. I have a fix for **openh264**, but it's not part of this PR. To enable this for other libraries, more work is required: like **openh264**, they are not configured out-of-the-box to run this code.

tanersener commented:
> This PR only takes care of libvpx. I have a fix for openh264, but it's not part of this PR. To enable this for other libraries, more work is required: like openh264, they are not configured out-of-the-box to run this code.

Yeah, I know. I'll test the latest changes. If everything goes well then we can merge this PR.

tanersener commented:
Tested the latest updates and didn't receive any errors. Thanks for resolving all the code review items.

alexcohn commented:
>  I have a fix for openh264, but it's not part of this PR

So, in the end, they [are](https://github.com/tanersener/mobile-ffmpeg/commit/888d62266255ee50f8f3bd3b818fbb8776241f9b#diff-0cc4895ba8de3b8c4b31e0583c0d580f). I hope it's for good.

tanersener commented:
I guess it will be fine. If there is something that we missed during the tests it will come out later and we can find a fix for it. 

The important thing is `mobile-ffmpeg` won't use the deprecated `cpu-features` library anymore. 

Another important step is taken by using a submodule for one the libraries. If it works well then most of the external libraries will be updated to use submodules and repository size will decrease. But more time is needed for that. There are more important issues I need to concentrate on right now.

tanersener commented:
Made some tests on submodules to see whether I can switch to submodules for all possible external libraries. Updated two of the external libraries to use submodules from github. Played a little bit with `--depth` and `shallow` options. 

Pros:
- Initial repository size decreased if I don't clone the project with submodules
- Can be configured to update/clone a submodule only when that external library is enabled. So you won't download the source code of a external library that you won't build

Cons:
- After cloning a submodule I noticed that the total size of the repository increases. `.git/modules` directory is consuming too much space. More than the source code of an eternal library. `--depth` and `shallow` options help a little bit doesn't solve the problem

I believe that for repositories where context or history of the external library is important, submodules can be helpful. Additionally, there are some positives that `mobile-ffmpeg` can benefit by switching to `submodules` too. Like not downloading the source code of unused external libraries. However, the main problem is still there. Repository is still very big. And it becomes even bigger for most of the cases.

alexcohn commented:
An alternative to submodules which reduces the total repo size, could be something like https://gerrit.googlesource.com/git-repo/. Still, I find submodules work nicely, especially when both the host and the submodule are hosted on GitHub. I understand your concern that `.git` folder grows much more with submodules, but as you write, this does not happen on the first `clone`, only limited to the subprojects that are used. And the possibility to lookup history of some 'external' library file is invaluable.

tanersener commented:
Thanks for the advice. I'm not sure if installing another tool is a good idea for this. If it had been possible to download a commit as a zip file without a `.git` folder inside that would be ideal in this case. But we don't have that option. I think I'll create a few bash functions to do something similar. `git clone` a specific commit from the repo and remove `.git` folder for them. Modifying `.git` folder of the parent project is risky but removing the `.git` folder of the external library shouldn't be a problem.

alexcohn commented:
For projects hosted on GitHub, you can download a zip without `.git` for any chosen commit (you probably want to fork the original to be sure that the relevant hash does not go away). Even `x265` and `ffmpeg` have official mirrors on GitHub. As far as I know, `xvidcore` is 

If you think it's worth it, I could take this task on myself, including the feature that will only download the repos for enabled libs. You still have to fork all the libs (but luckily, you can do it entirely on the Web).

tanersener commented:
Thanks, I already pushed something about this on the `development` branch yesterday, https://github.com/tanersener/mobile-ffmpeg/commit/f6189815f006b5f5dd10157ca059adc8fb1956ae. Moved duplicate functions into an `arch-common.sh` file and added a `clone_git_repository` function to clone the source code. Currently `cpu-features`, `expat` and `openh264` uses this. I'm not happy about the implementation though. There are some unnecessary steps there. If you have ideas to improve it, like downloading a zip file without git information, I'll be happy to merge it to the `development`. 

It can be extended to cover other libraries as well but first I want to make sure that this is the right way to go.

alexcohn commented:
Good, maybe it's worth to create a new issue for that?

tanersener commented:
Yeah, created this ticket, https://github.com/tanersener/mobile-ffmpeg/issues/459.
